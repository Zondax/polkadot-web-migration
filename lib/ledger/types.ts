import Transport from '@ledgerhq/hw-transport'
import { PolkadotGenericApp } from '@zondax/ledger-substrate'

/**
 * Represents a response object from a connection request.
 */
export interface ConnectionResponse {
  error?: string
  connection?: DeviceConnectionProps
}

/**
 * Represents the properties of a device connection.
 */
export interface DeviceConnectionProps {
  transport?: Transport
  genericApp?: PolkadotGenericApp
  isAppOpen: boolean
}
