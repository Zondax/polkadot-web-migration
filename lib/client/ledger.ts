// Re-export ledger client from state
export { ledgerClient } from 'state/client/ledger'
