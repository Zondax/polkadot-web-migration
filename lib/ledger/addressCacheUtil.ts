import { encodeAddress } from '@polkadot/keyring'
import type { GenericeResponseAddress } from '@zondax/ledger-substrate/dist/common'

interface CachedPublicKey {
  publicKey: Uint8Array // Raw public key from Ledger
}

/**
 * Address cache utility for Ledger public keys
 *
 * This cache stores public keys by BIP44 derivation path and generates
 * SS58-formatted addresses on demand. This eliminates redundant Ledger
 * device calls when scanning multiple blockchain applications that use
 * the same derivation paths but different SS58 prefixes.
 */
class AddressCache {
  private cache = new Map<string, CachedPublicKey>()

  /**
   * Check if an address is cached for the given path
   * @param path BIP44 derivation path
   * @returns true if cached, false otherwise
   */
  has(path: string): boolean {
    return this.cache.has(path)
  }

  /**
   * Get cached address with specific SS58 prefix
   * @param path BIP44 derivation path
   * @param ss58Prefix SS58 format prefix for the target network
   * @returns GenericeResponseAddress if cached, null if not found
   */
  get(path: string, ss58Prefix: number): GenericeResponseAddress | null {
    const cached = this.cache.get(path)
    if (!cached) return null

    // Generate address with requested SS58 prefix using cached public key
    const address = encodeAddress(cached.publicKey, ss58Prefix)
    // Convert public key back to hex string format expected by GenericeResponseAddress
    const pubKeyHex = Buffer.from(cached.publicKey).toString('hex')
    return {
      address,
      pubKey: pubKeyHex,
      path,
    } as GenericeResponseAddress
  }

  /**
   * Store public key for the given path
   * @param path BIP44 derivation path
   * @param publicKey Raw public key from Ledger device
   */
  set(path: string, publicKey: Uint8Array): void {
    this.cache.set(path, { publicKey })
  }

  /**
   * Clear entire cache
   */
  clear(): void {
    this.cache.clear()
  }

  /**
   * Remove specific cache entry
   * @param path BIP44 derivation path to remove
   */
  remove(path: string): void {
    this.cache.delete(path)
  }

  /**
   * Get cache statistics
   * @returns Object containing cache size and paths
   */
  getStats() {
    return {
      size: this.cache.size,
      paths: Array.from(this.cache.keys()),
    }
  }
}

// Export singleton instance
export const addressCache = new AddressCache()
