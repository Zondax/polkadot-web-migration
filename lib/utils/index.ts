// Re-export all utilities for backwards compatibility
export * from './actions'
export * from './address'
export * from './balance'
export * from './browser'
export * from './clipboard'
export * from './error'
export * from './format'
export * from './html'
export * from './ledger'
export * from './nft'
export * from './styles'
