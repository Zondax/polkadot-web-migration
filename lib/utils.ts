import { ResponseVersion } from '@zondax/ledger-js';
import axios from 'axios';
import { type ClassValue, clsx } from 'clsx';
import {
  errorDetails,
  InternalErrors,
  LedgerErrorDetails,
  LedgerErrors
} from 'config/errors';
import { LedgerClientError } from 'state/client/base';
import { notifications$ } from 'state/notifications';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function handleLedgerError(
  error: LedgerClientError,
  defaultError: InternalErrors | LedgerErrors
): LedgerErrorDetails {
  let resolvedError: LedgerErrorDetails | undefined;

  const errorDetail = errorDetails[error.name as keyof typeof errorDetails];
  if (errorDetail) {
    resolvedError = errorDetail;
  } else {
    resolvedError = errorDetails[defaultError] || errorDetails.default;
  }

  notifications$.push({
    title: resolvedError.title,
    description: resolvedError.description ?? '',
    type: 'error',
    autoHideDuration: 5000
  });

  return resolvedError;
}

/**
 * Copies a string to the clipboard.
 * @param text - The text to copy.
 * @throws Will throw an error if the copy operation fails.
 */
export const copyContent = async (text: string) => {
  try {
    await navigator.clipboard.writeText(text.toString());
  } catch (err) {
    console.log('Failed to copy content ', err);
    throw err;
  }
};

/**
 * Truncates the middle of a string to a specified maximum length.
 * @param str - The string to truncate.
 * @param maxLength - The maximum length of the string.
 * @returns The truncated string, or undefined if the input string is empty.
 */
export const truncateMiddleOfString = (str: string, maxLength: number) => {
  if (!str) {
    return null;
  }
  if (str.length <= maxLength) {
    return str;
  }
  const middle = Math.floor(maxLength / 2);
  const start = str.substring(0, middle);
  const end = str.substring(str.length - middle, str.length);
  return `${start}...${end}`;
};

/**
 * Formats a balance to a human-readable string.
 *
 * @param {number} balance - The balance to format.
 * @returns {string} The formatted balance.
 */
export const formatBalance = (
  balance: number,
  ticker: string,
  decimals?: number
): string => {
  if (balance === 0) {
    return `0 ${ticker}`;
  }

  const adjustedBalance = decimals ? balance / Math.pow(10, decimals) : balance;

  const formattedBalance = adjustedBalance.toLocaleString(undefined, {
    minimumFractionDigits: 0,
    maximumFractionDigits: 5
  });

  return `${formattedBalance} ${ticker}`;
};

export const formatVersion = (version: ResponseVersion): string => {
  const { major, minor, patch } = version;
  return `${major}.${minor}.${patch}`;
};

export const getAppLightIcon = async (appId: string) => {
  try {
    const hubUrl = process.env.NEXT_PUBLIC_HUB_BACKEND_URL;

    if (!hubUrl) {
      return;
    }

    const response = await axios.get(hubUrl + `/app/${appId}/icon/light`);
    return { data: response.data, error: undefined };
  } catch (error) {
    return { data: [], error: 'error' };
  }
};
