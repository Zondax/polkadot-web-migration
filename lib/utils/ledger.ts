import axios from 'axios'
import { type App, AppStatus } from 'state/ledger'
import {
  type Address,
  type AddressBalance,
  type AddressWithVerificationStatus,
  type MultisigAddress,
  VerificationStatus,
} from 'state/types/ledger'
import { hasAddressBalance, hasBalance } from './balance'

/**
 * Retrieves a light icon for a given app from the Hub backend.
 *
 * @param appId - The ID of the app to retrieve the icon for.
 * @returns The icon data and any error that occurred.
 */
export const getAppLightIcon = async (appId: string) => {
  try {
    // First try to fetch from API
    const hubUrl = process.env.NEXT_PUBLIC_HUB_BACKEND_URL

    if (!hubUrl) {
      return { data: undefined, error: 'Hub URL not configured' }
    }

    try {
      const response = await axios.get(`${hubUrl}/app/${appId}/icon/light`)
      return { data: response.data, error: undefined }
    } catch (_apiError) {
      // API call failed, try local image as fallback
    }

    // If API fetch fails, check if the image exists locally
    const localImagePath = `/logos/chains/${appId}.svg`

    try {
      const res = await fetch(localImagePath)
      if (res.ok) {
        // For SVG files, we need to get the text content
        const svgContent = await res.text()
        return { data: svgContent, error: undefined }
      }
    } catch (_localError) {
      // Local image doesn't exist either
    }

    // If we get here, both API and local fetches failed
    return { data: undefined, error: 'Icon not found' }
  } catch (_error) {
    return { data: undefined, error: 'Error fetching app icon' }
  }
}

/**
 * Filters apps to only include those that were validly synchronized and have balances.
 *
 * @param apps - The apps to filter.
 * @returns Apps that are validly synchronized and have balances.
 */
export const filterValidSyncedAppsWithBalances = (apps: App[]): App[] => {
  return apps
    .map(app => ({
      ...app,
      accounts:
        app.accounts?.filter(
          (account: Address) => (!account.error || account.error?.source === 'migration') && hasAddressBalance(account)
        ) || [],
      multisigAccounts:
        app.multisigAccounts?.filter(
          (account: MultisigAddress) => (!account.error || account.error?.source === 'migration') && hasAddressBalance(account)
        ) || [],
    }))
    .filter(app => app.accounts.length > 0 || app.multisigAccounts?.length > 0)
}

/**
 * Checks if a single account meets ALL criteria for migration.
 * This is the single source of truth for migration eligibility.
 *
 * @param account - The account to check (Address or MultisigAddress)
 * @param checkSelected - Whether to check if account is selected (default: true)
 * @returns true if the account can be migrated, false otherwise
 */
export const canAccountBeMigrated = (account: Address | MultisigAddress) => {
  // Must be selected (unless explicitly skipped)
  if (!account.selected) return false

  // Must not be already migrated
  if (account.status === 'migrated') return false

  // Must have balances
  if (!account.balances || account.balances.length === 0) return false

  // Allow accounts with migration errors, but not other errors
  if (account.error && account.error.source !== 'migration') return false

  // ALL balances must meet migration criteria
  return account.balances.every(balance => {
    // Must have transferable balance
    if (!hasBalance([balance], true)) return false

    // Must have destination address - THIS IS THE KEY CHECK!
    if (!balance.transaction?.destinationAddress) return false

    return true
  })
}

/**
 * Filters apps to only include those that have accounts that can be migrated.
 * This function filters for accounts that are selected and have a balance and destination address.
 *
 * @param apps - The apps to filter.
 * @returns Apps with accounts that can be migrated.
 */
export const filterValidSelectedAccountsForMigration = (apps: App[]): App[] => {
  const filteredApps = apps
    .map(app => ({
      ...app,
      accounts: app.accounts?.filter(account => canAccountBeMigrated(account)) || [],
      multisigAccounts: app.multisigAccounts?.filter(account => canAccountBeMigrated(account)) || [],
    }))
    .filter(app => app.accounts.length > 0 || app.multisigAccounts?.length > 0)

  return filteredApps
}

/**
 * Filters apps to include only those with accounts or multisig accounts that have errors (excluding migration errors).
 *
 * @param apps - The list of apps to filter.
 * @returns Apps containing accounts or multisig accounts with non-migration errors.
 */
export const filterInvalidSyncedApps = (apps: App[]): App[] => {
  return apps
    .map(app => ({
      ...app,
      accounts: app.accounts?.filter((account: Address) => account.error && account.error?.source !== 'migration') || [],
      multisigAccounts:
        app.multisigAccounts?.filter((account: MultisigAddress) => account.error && account.error?.source !== 'migration') || [],
    }))
    .filter(app => app.accounts.length > 0 || app.multisigAccounts?.length > 0 || app.status === 'error')
}

/**
 * Checks if there are any accounts with errors.
 *
 * @param apps - The apps to check.
 * @returns True if there are accounts with errors, false otherwise.
 */
export const hasAccountsWithErrors = (apps: App[]): boolean => {
  return apps.some(
    app =>
      app.error?.source === 'synchronization' ||
      app.status === AppStatus.RESCANNING ||
      app.accounts?.some(account => account.error && account.error?.source !== 'migration') ||
      app.multisigAccounts?.some(account => account.error && account.error?.source !== 'migration')
  )
}

/**
 * Checks if the app has any accounts.
 *
 * @param app - The app to check.
 * @returns True if the app has accounts, false otherwise.
 */
export const hasAppAccounts = (app: App): boolean => {
  return Boolean((app.accounts && app.accounts.length > 0) || (app.multisigAccounts && app.multisigAccounts.length > 0))
}

/**
 * Gets the total number of accounts with balances for a single app.
 *
 * @param app - The app to check.
 * @returns The number of accounts with balances for the app.
 */
export const getAppTotalAccounts = (app: App): number => {
  return (app.accounts?.length || 0) + (app.multisigAccounts?.length || 0)
}

/**
 * Filters the accounts that will be saved in the app.
 *
 * @param app - The app whose accounts will be filtered.
 * @param filterByBalance - A function to test each account. Return true to keep the account, false otherwise.
 * @returns An array of accounts that satisfy the predicate.
 */
export function filterAccountsForApps<T extends Address | MultisigAddress>(accounts: T[], filterByBalance: boolean): T[] {
  return accounts.filter(
    account =>
      !filterByBalance ||
      (account.balances && hasBalance(account.balances)) ||
      account.error ||
      (account.memberMultisigAddresses && account.memberMultisigAddresses.length > 0)
  )
}

/**
 * Sets a default destination address for all balances in an account that have transactions
 * @param account The account (Address or MultisigAddress) to update
 * @param defaultDestinationAddress The default destination address to set
 * @returns The account with updated transaction destination addresses
 */
export function setDefaultDestinationAddress<T extends { balances?: AddressBalance[] }>(account: T, defaultDestinationAddress: string): T {
  if (!account.balances) {
    return account
  }

  return {
    ...account,
    balances: account.balances.map(balance => ({
      ...balance,
      transaction: {
        ...balance.transaction,
        destinationAddress: balance.transaction?.destinationAddress || defaultDestinationAddress,
      },
    })),
  }
}

/**
 * Adds destination addresses from a list of accounts to an address map
 * @param accounts - The accounts to process (can be regular or multisig accounts)
 * @param addressMap - The map to store unique addresses with their paths and status
 * @param polkadotAddresses - The list of polkadot addresses to check against
 */
export function addDestinationAddressesFromAccounts(
  accounts: Address[] | MultisigAddress[] | undefined,
  addressMap: Map<string, AddressWithVerificationStatus>,
  polkadotAddresses: string[]
): void {
  if (!accounts) return
  for (const account of accounts) {
    if (account.balances && account.balances.length > 0) {
      for (const balance of account.balances) {
        if (
          balance.transaction?.destinationAddress &&
          !addressMap.has(balance.transaction.destinationAddress) &&
          polkadotAddresses.includes(balance.transaction.destinationAddress)
        ) {
          addressMap.set(balance.transaction.destinationAddress, {
            address: balance.transaction.destinationAddress,
            path: account.path,
            status: VerificationStatus.PENDING,
          })
        }
      }
    }
  }
}
