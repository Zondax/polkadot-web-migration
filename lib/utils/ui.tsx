import { BN } from '@polkadot/util'
import { AlertCircle, AtSign, CheckCircle, Clock, Globe, Mail, Twitter, User, Users, XCircle } from 'lucide-react'

import type { TooltipItem } from '@/components/CustomTooltip'
import { ExplorerLink } from '@/components/ExplorerLink'
import { Spinner } from '@/components/icons'
import type { AppId, Token } from '@/config/apps'
import { ExplorerItemType } from '@/config/explorers'
import { type Registration, TransactionStatus } from '@/state/types/ledger'
import { formatBalance } from './format'

/**
 * Returns a status icon and message corresponding to the given transaction status.
 *
 * @param status - The current transaction status (optional).
 * @param txStatusMessage - An optional custom status message to display.
 * @param size - The size of the status icon ('sm', 'md', or 'lg'). Defaults to 'sm'.
 * @param txHash - Optional transaction hash for explorer links.
 * @returns An object containing the statusIcon (ReactNode), an optional statusMessage (string), and the txHash if available.
 */
export const getTransactionStatus = (
  status?: TransactionStatus,
  txStatusMessage?: string,
  size: 'sm' | 'md' | 'lg' = 'sm',
  txHash?: string
): { statusIcon: React.ReactNode; statusMessage?: string; txHash?: string } => {
  let statusIcon: React.ReactNode | null = null
  let statusMessage = txStatusMessage

  const sizeClasses = {
    sm: 'h-4 w-4',
    md: 'h-6 w-6',
    lg: 'h-8 w-8',
  }

  const iconSize = sizeClasses[size]

  switch (status) {
    case TransactionStatus.IS_LOADING:
      statusIcon = <Spinner />
      statusMessage = 'Loading...'
      break
    case TransactionStatus.PREPARING_TX:
      statusIcon = <Spinner />
      statusMessage = 'Preparing transaction...'
      break
    case TransactionStatus.SIGNING:
      statusIcon = <Spinner />
      statusMessage = 'Please sign the transaction in your Ledger device'
      break
    case TransactionStatus.SUBMITTING:
      statusIcon = <Spinner />
      statusMessage = 'Submitting transaction...'
      break
    case TransactionStatus.PENDING:
      statusIcon = <Clock className={`${iconSize} text-muted-foreground`} />
      statusMessage = 'Transaction pending...'
      break
    case TransactionStatus.IN_BLOCK:
      statusIcon = <Clock className={`${iconSize} text-muted-foreground`} />
      break
    case TransactionStatus.FINALIZED:
      statusIcon = <Clock className={`${iconSize} text-muted-foreground`} />
      break
    case TransactionStatus.SUCCESS:
      statusIcon = <CheckCircle className={`${iconSize} text-green-500`} />
      break
    case TransactionStatus.FAILED:
      statusIcon = <XCircle className={`${iconSize} text-red-500`} />
      break
    case TransactionStatus.ERROR:
      statusIcon = <AlertCircle className={`${iconSize} text-red-500`} />
      break
    case TransactionStatus.WARNING:
      statusIcon = <AlertCircle className={`${iconSize} text-yellow-500`} />
      break
    case TransactionStatus.COMPLETED:
      statusIcon = <Clock className={`${iconSize} text-muted-foreground`} />
      break
    default:
      statusIcon = (
        <span className="px-2 py-1 text-xs rounded-full bg-polkadot-lime text-black border border-storm-200">Ready to migrate</span>
      )
  }
  return { statusIcon, statusMessage, txHash }
}

/**
 * Validates a numeric input value against required, numeric, positive, and maximum constraints.
 *
 * @param value - The input value as a string to validate.
 * @param max - The maximum allowed value (inclusive).
 * @returns An object with a boolean `valid` flag and a `helperText` message for user feedback.
 */
export const validateNumberInput = (value: number, max: BN, token: Token): { valid: boolean; helperText: string } => {
  if (Number.isNaN(value)) {
    return { valid: false, helperText: 'Amount is required.' }
  }
  // Convert value to BN for comparison
  let inputBN: BN
  try {
    inputBN = new BN(value)
  } catch (_e) {
    return { valid: false, helperText: 'Amount must be a number.' }
  }
  if (inputBN.lte(new BN(0))) {
    return { valid: false, helperText: 'Amount must be greater than zero.' }
  }
  if (inputBN.gt(max)) {
    return { valid: false, helperText: `Amount cannot exceed your staked balance (${formatBalance(max, token, token?.decimals)}).` }
  }
  return { valid: true, helperText: '' }
}

/**
 * Returns an array of identity items for display from a Registration object.
 * Each item contains a label, value, icon, and optional href.
 */
export function getIdentityItems(registration: Registration | undefined, appId: AppId): TooltipItem[] {
  if (!registration?.identity) return []
  const identity = registration.identity
  const items = [
    {
      label: 'Parent account',
      value: identity.parent ? (
        <ExplorerLink
          value={identity.parent}
          appId={appId}
          explorerLinkType={ExplorerItemType.Address}
          truncate={false}
          disableTooltip
          className="break-all"
          size="xs"
        />
      ) : undefined,
      icon: Users,
    },
    {
      label: 'Parent legal name',
      value: identity.displayParent,
      icon: AtSign,
    },
    {
      label: 'Display name',
      value: identity.display,
      icon: User,
    },
    {
      label: 'Legal name',
      value: identity.legal,
      icon: AtSign,
    },
    {
      label: 'Website',
      value: identity.web,
      icon: Globe,
      href: identity.web,
    },
    {
      label: 'Email',
      value: identity.email,
      icon: Mail,
      href: identity.email ? `mailto:${identity.email}` : undefined,
    },
    {
      label: 'Twitter',
      value: identity.twitter,
      icon: Twitter,
      href: identity.twitter,
    },
  ]
  return items.filter(item => item.value !== undefined) as TooltipItem[]
}
