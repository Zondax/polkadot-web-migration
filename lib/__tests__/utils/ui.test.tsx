import { TransactionStatus } from 'state/types/ledger'
import { describe, expect, it, vi } from 'vitest'

import { BN } from '@polkadot/util'
import { getTransactionStatus, validateNumberInput } from '../../utils/ui'
import { mockApp1 } from './__mocks__/mockData'

// Mock lucide-react icons and Spinner
vi.mock('lucide-react', () => ({
  AlertCircle: (props: any) => <div data-testid="alert-circle" {...props} />,
  CheckCircle: (props: any) => <div data-testid="check-circle" {...props} />,
  Clock: (props: any) => <div data-testid="clock" {...props} />,
  XCircle: (props: any) => <div data-testid="x-circle" {...props} />,
}))
vi.mock('@/components/icons', () => ({
  Spinner: () => <div data-testid="spinner" />,
}))

describe('getTransactionStatus', () => {
  it('returns Spinner and message for IS_LOADING', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.IS_LOADING)
    expect(statusMessage).toBe('Loading...')
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns Clock and message for PENDING', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.PENDING)
    expect(statusMessage).toBe('Transaction pending...')
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns Clock for IN_BLOCK', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.IN_BLOCK)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns Clock for FINALIZED', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.FINALIZED)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns CheckCircle for SUCCESS', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.SUCCESS)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns XCircle for FAILED', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.FAILED)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns AlertCircle for ERROR', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.ERROR)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns AlertCircle for WARNING', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.WARNING)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns Clock for COMPLETED', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(TransactionStatus.COMPLETED)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns default for undefined status', () => {
    const { statusIcon, statusMessage } = getTransactionStatus(undefined)
    expect(statusMessage).toBeUndefined()
    expect(statusIcon).toMatchSnapshot()
  })
  it('returns custom message if provided', () => {
    const { statusMessage } = getTransactionStatus(TransactionStatus.SUCCESS, 'Custom!')
    expect(statusMessage).toBe('Custom!')
  })
  it('applies correct icon size', () => {
    const { statusIcon } = getTransactionStatus(TransactionStatus.SUCCESS, undefined, 'lg')
    expect(statusIcon).toMatchSnapshot()
  })
})

describe('validateNumberInput', () => {
  const max = new BN(1000)
  const mockToken = mockApp1.token
  it('returns valid: false and required message for NaN', () => {
    const result = validateNumberInput(Number.NaN, max, mockToken)
    expect(result.valid).toBe(false)
    expect(result.helperText).toBe('Amount is required.')
  })
  it('returns valid: false and number message for non-numeric', () => {
    const result = validateNumberInput('abc' as unknown as number, max, mockToken)
    expect(result.valid).toBe(false)
    expect(result.helperText).toBe('Amount must be a number.')
  })
  it('returns valid: false and greater than zero message for zero or negative', () => {
    expect(validateNumberInput(0, max, mockToken)).toEqual({ valid: false, helperText: 'Amount must be greater than zero.' })
    expect(validateNumberInput(-5, max, mockToken)).toEqual({ valid: false, helperText: 'Amount must be greater than zero.' })
  })
  it('returns valid: false and max message for value above max', () => {
    const result = validateNumberInput(1001, max, mockToken)
    expect(result.valid).toBe(false)
    expect(result.helperText).toBe('Amount cannot exceed your staked balance (0.0000001 DOT).')
  })
  it('returns valid: true for valid numbers within range', () => {
    expect(validateNumberInput(123, max, mockToken)).toEqual({ valid: true, helperText: '' })
    expect(validateNumberInput(123.45, max, mockToken)).toEqual({ valid: true, helperText: '' })
    expect(validateNumberInput(1000, max, mockToken)).toEqual({ valid: true, helperText: '' })
  })
})
