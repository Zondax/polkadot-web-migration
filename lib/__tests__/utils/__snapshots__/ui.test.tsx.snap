// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTransactionStatus > applies correct icon size 1`] = `
<CheckCircle
  className="h-8 w-8 text-green-500"
/>
`;

exports[`getTransactionStatus > returns AlertCircle for ERROR 1`] = `
<AlertCircle
  className="h-4 w-4 text-red-500"
/>
`;

exports[`getTransactionStatus > returns AlertCircle for WARNING 1`] = `
<AlertCircle
  className="h-4 w-4 text-yellow-500"
/>
`;

exports[`getTransactionStatus > returns CheckCircle for SUCCESS 1`] = `
<CheckCircle
  className="h-4 w-4 text-green-500"
/>
`;

exports[`getTransactionStatus > returns Clock and message for PENDING 1`] = `
<Clock
  className="h-4 w-4 text-muted-foreground"
/>
`;

exports[`getTransactionStatus > returns Clock for COMPLETED 1`] = `
<Clock
  className="h-4 w-4 text-muted-foreground"
/>
`;

exports[`getTransactionStatus > returns Clock for FINALIZED 1`] = `
<Clock
  className="h-4 w-4 text-muted-foreground"
/>
`;

exports[`getTransactionStatus > returns Clock for IN_BLOCK 1`] = `
<Clock
  className="h-4 w-4 text-muted-foreground"
/>
`;

exports[`getTransactionStatus > returns Spinner and message for IS_LOADING 1`] = `<Spinner />`;

exports[`getTransactionStatus > returns XCircle for FAILED 1`] = `
<XCircle
  className="h-4 w-4 text-red-500"
/>
`;

exports[`getTransactionStatus > returns default for undefined status 1`] = `
<span
  className="px-2 py-1 text-xs rounded-full bg-polkadot-lime text-black border border-storm-200"
>
  Ready to migrate
</span>
`;
